File: Templates/Graphs/GraphAdjList.cpp
//
// Code Contribution(s): Rajeev Ranjan
//

#include <iostream>

using namespace std;

class Node
{
	public:
		int data;
		Node * next;
};

class LinkedList
{
	public:
		Node * insertNode(Node * currentHead, int vertex);
};

Node * LinkedList::insertNode(Node * currentHead, int vertex)
{
	Node * newNode = new Node;

	newNode->data = vertex;
	newNode->next = currentHead;

	return newNode;
}

int main()
{
	LinkedList l;
	int v;
	cin>>v;
	int e1, e2;
	Node * adjList[v+1]={NULL};
	for(int i = 1; i<v; i++)
	{
		cin>>e1>>e2;
		adjList[e1] = l.insertNode(adjList[e1], e2);
		adjList[e2] = l.insertNode(adjList[e2], e1);
	}
	for(int i = 1; i<=v; i++)
	{
		Node * traverse = adjList[i];
		cout<<i<<"->";
		while(traverse->next != NULL)
		{
			cout<<traverse->data<<"->";
			traverse = traverse->next;
		}
		cout<<traverse->data<<"->NULL"<<endl;
	}
	return 0;
}

File: Implementations/Python/Graph/GraphAdjList.py
#
# Code Contributor(s) : Rajeev Ranjan
#

import sys


class Node:

	def __init__(self):
		self.data = 0
		self.next = None

def makeNode(current, data):

	newNode = Node()
	newNode.data = data
	newNode.next = current

	return newNode

if __name__ == '__main__':

	numberOfNodes = int(input('Enter number of Nodes : '))
	numberOfEdges = int(input('Enter number of Edges : '))

	graph = [None] * (numberOfNodes+1)

	for i in range(numberOfEdges):
		start = int(input())
		end = int(input())
		graph[start] = makeNode(graph[start], end)
		graph[end] = makeNode(graph[end], start)



	for i in range(numberOfNodes+1):
		temp = graph[i]
		print(i, end='->')
		while temp is not None:
			print(temp.data,end='->')
			temp = temp.next
		print('NULL')
